Allow trailing newline in tokenFile

benburry opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**:

When referencing a tokenFile in your kubeconfig user entry (e.g. to be used by
kubectl), the contents of this token file are sent verbatim as the `Bearer` http
header. This causes issues if the tokenFile contains trailing (or leading)
whitespace, such as a single linebreak at the end of the file.

`Unable to connect to the server: net/http: invalid header field value`

It would be reasonable to ask people to not include any leading/trailing
whitespace in their token files. However, some (many? most?) text editors will
add a single linebreak at the end of the line by default. 

This commit removes any leading or trailing whitespace from the token read from
tokenFile, before trying to use it as the bearer token header. 

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #52919

**Special notes for your reviewer**:

I've added a single test to validate the change: `TestTrailingLinebreakTokenFile`

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
Allow trailing newline in tokenFile
```

@kubernetes/sig-cli-pr-reviews

benburry committed:
Be permissive of whitespace in tokenFile

When referencing a tokenFile in your kubeconfig user entry (e.g. to be used by
kubectl), the contents of this token file are sent verbatim as the Bearer
header. This causes issues if the tokenFile contains trailing (or leading)
whitespace, such as a single linebreak at the end of the file.

`Unable to connect to the server: net/http: invalid header field value`

It would be reasonable to ask people to not include any leading/trailing
whitespace in their token files. However, some (many? most?) text editors will
add a single linebreak at the end of the line by default. This seems to
originate from the POSIX definition of Line:

`A sequence of zero or more non- <newline> characters plus a terminating
<newline> character.` (from
http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_206)

This commit removes any leading or trailing whitespace from the token read from
tokenFile, before trying to use it as the bearer token header. I've added a
single test to validate the change.

liggitt commented:
this seems like a helpful thing to do, but we have to be careful about requiring transformation of info in kubeconfig files... clients in other languages are expected to be able to consume kubeconfig files as well... will we require them to mimic all the normalizations client-go does?

dims commented:
/ok-to-test

benburry commented:
/retest

lavalamp commented:
I can sorta buy removing a trailing newline, as we can define the meaning to be "the first line of the file". (Since bearer tokens can't have newlines in them anyway. [citation needed])

Removing an initial newline, or initial or trailing whitespace characters (as this code seems to do), seems well into the territory @liggitt is worried about and I'm not in favor of that.

benburry commented:
@lavalamp My thinking was along the same lines as yours - the token file is supposed to contain a Bearer token and nothing else. The spec (https://tools.ietf.org/html/rfc6750#section-2.1 ) defines this to be a single string with no newlines and no whitespace.

Perhaps (if the project is unwilling to accept the approach in this PR) a less contentious option would be to validate that the token file conforms to the bearer spec before sending it verbatim as an http header.

It seems that at least one or the other (strip whitespace or validate) would be wise.

liggitt commented:
> Perhaps (if the project is unwilling to accept the approach in this PR) a less contentious option would be to validate that the token file conforms to the bearer spec before sending it verbatim as an http header.

giving a clear error on an invalid token file seems better to me than fixing up data in one client, which encourages kubeconfig files that are not uniformly successful across clients

benburry committed:
Adjust tokenFile behaviour to read first line only

Read only the first line from tokenFile as the Bearer authorization token, as
the spec (rfc6750 section-2.1) defines it as not including newlines.

benburry commented:
@lavalamp @liggitt I have a new commit ready that uses bufio.Scanner to read only the first line from the tokenFile.
If the consensus is to prefer validation of tokenFile though, I can close this PR and work on the other approach separately.

ericchiang commented:
@benburry I agree with the other concerns raised. Producing a better error would probably be the better PR to send.

lavalamp commented:
Reading just the first line seems reasonable. (please push your commit so we can look)

Doing anything else seems wrong--if you add validation, it should fail rather than be normalized.

benburry commented:
@lavalamp I've pushed the code to just read the first line of the config file

lavalamp commented on staging/src/k8s.io/client-go/tools/clientcmd/client_config.go:
```diff
@@ -214,11 +215,21 @@ func (config *DirectClientConfig) getUserIdentificationPartialConfig(configAuthI
+
+		// read the first line (excluding newline) of the token file as
+		// the Bearer token
+		scanner := bufio.NewScanner(tokenFile)
```
Make a bufio.Reader and call ReadLine or ReadString("\n") instead of using a Scanner?

benburry replied:
I can certainly make that change, however for this edit I looked through the codebase to get a sense of the right approach:

There are many uses of Scanner in the codebase, but fewer to just read a single line. Of the two alternatives, ReadString seems to be the better approach (ReadLine would need a continually filled buffer, which seems to be reimplementing logic already in ReadString and Scanner), but still needs a second step to strip the trailing newline.

When I made this change, my thinking was that Scanner is the better option as it encapsulates all the logic needed, doesn't require additional changes to the returned string, and is already in use in the codebase to read a single line more commonly than ReadString.

I'm happy to make the change if there's a compelling reason I don't know about as to why ReadString is the better approach, and I can use this as a learning opportunity.

benburry replied:
@lavalamp if you give me the nod that changing the code to use ReadString is a requirement for the PR to be accepted, I'm happy to make that change

benburry commented:
Are we waiting on more changes from me before moving this PR forward?
If there's not consensus on the change I'm happy to close this and write a tokenFile validation in its place.

ericchiang commented:
I'm with @liggitt on this. kubectl already has a lot of magic behavior that other clients struggle to replicate https://github.com/kubernetes-client

I don't think we should add this.

benburry commented:
Sure, that seems fair enough. I'll close this PR and put some validation in for the tokenfile instead
